File: codetraverse/ast_diff/resdiffer.py
import subprocess
import json
import re
from collections import defaultdict
from tree_sitter import Language, Parser, Node
import tree_sitter_rescript
import hashlib
import os
from .Detailedchanges import DetailedChanges
from .basefilediff import BaseFileDiff

def format_rescript_file(file_pth):
    try:
        subprocess.run(["npx", "rescript", "format", file_pth], capture_output=True)
    except:
        pass


class RescriptFileDiff(BaseFileDiff):
    def __init__(self, module_name=""):
        self.changes = DetailedChanges(module_name)

    def get_decl_name(self, node: Node, node_type: str, name_type: str) -> str:
        for child in node.children:
            if node_type and child.type == node_type:
                for grandchild in child.children:
                    if grandchild.is_named and grandchild.type == name_type:
                        return grandchild.text.decode(errors="ignore")
            elif not node_type and child.is_named and child.type == name_type:
                return child.text.decode(errors="ignore")
        return None

    def deep_equal(self, nodeA: Node, nodeB: Node):
        if (nodeA is None) != (nodeB is None):
            return False

        if nodeA is None and nodeB is None:
            return True

        if nodeA.type != nodeB.type:
            return False

        childrenA = nodeA.children
        childrenB = nodeB.children

        if len(childrenA) != len(childrenB):
            return False

        if len(childrenA) == 0:
            return nodeA.text == nodeB.text and nodeA.parent.text == nodeB.parent.text

        for childA, childB in zip(childrenA, childrenB):
            if not self.deep_equal(childA, childB):
                return False

        return True

    def extract_components(self, root: Node):
        queue = [root]
        
        functions = {}
        types = {}
        externals = {}

        node_name_mapper = {
            "let_declaration": (functions, lambda x: self.get_decl_name(x, "let_binding", "value_identifier")),
            "type_declaration": (types, lambda x: self.get_decl_name(x, "type_binding", "type_identifier")),
            "external_declaration": (externals, lambda x: self.get_decl_name(x, None, "value_identifier"))
        }

        while queue:
            current_node = queue.pop()
            if current_node.type in node_name_mapper.keys():
                dct, mapper_function = node_name_mapper[current_node.type]
                name = mapper_function(current_node)
                if name:
                    if current_node.parent.type != "source_file":
                        try:
                            name = f"{current_node.parent.parent.child(0).text.decode()}::{name}"
                        except:
                            pass
                    dct[name] = (current_node, current_node.text.decode(errors="ignore"), current_node.start_point, current_node.end_point)
            else:
                for child in reversed(current_node.children):
                    if child.is_named:
                        queue.append(child)
        return functions, types, externals

    def diff_components(self, before_map: dict, after_map: dict) -> dict:
        before_names = set(before_map.keys())
        after_names = set(after_map.keys())

        added_names = after_names - before_names
        deleted_names = before_names - after_names
        common = before_names & after_names

        added = [(n, after_map[n][1], {"start": after_map[n][2], "end": after_map[n][3]}) for n in sorted(added_names)]
        deleted = [(n, before_map[n][1], {"start": before_map[n][2], "end": before_map[n][3]}) for n in sorted(deleted_names)]

        modified = []
        for name in sorted(common):
            old_ast, old_body, old_start, old_end = before_map[name]
            new_ast, new_body, new_start, new_end = after_map[name]
            if not self.deep_equal(old_ast, new_ast):
                modified.append((name, old_body, new_body, {"old_start": old_start, "old_end": old_end, "new_start": new_start, "new_end": new_end}))

        return {"added": added, "deleted": deleted, "modified": modified}

    def compare_two_files(self, old_file_ast, new_file_ast) -> DetailedChanges:
        """The main method to compare two ReScript files."""
        old_funcs, old_types, old_ext = self.extract_components(old_file_ast.root_node)
        new_funcs, new_types, new_ext = self.extract_components(new_file_ast.root_node)

        category_map = {
            "functions": (old_funcs, new_funcs),
            "types": (old_types, new_types),
            "externals": (old_ext, new_ext),
        }

        for category, (old_map, new_map) in category_map.items():
            diff = self.diff_components(old_map, new_map)
            for change_type in ["added", "deleted", "modified"]:
                for item in diff[change_type]:
                    self.changes.add_change(category, change_type, item)

        return self.changes


    def process_single_file(self, file_ast, mode="deleted"):
        funcs, types, exts = self.extract_components(file_ast.root_node)
        func_names = set(funcs.keys())
        type_names = set(types.keys())
        ext_names = set(exts.keys())

        if mode == "deleted":
            self.changes.deletedFunctions = [(n, funcs[n][1], {"start": funcs[n][2], "end": funcs[n][3]}) for n in sorted(func_names)]
            self.changes.deletedTypes = [(n, types[n][1], {"start": types[n][2], "end": types[n][3]}) for n in sorted(type_names)]
            self.changes.deletedExternals = [(n, exts[n][1], {"start": exts[n][2], "end": exts[n][3]}) for n in sorted(ext_names)]
        else:
            self.changes.addedFunctions = [(n, funcs[n][1], {"start": funcs[n][2], "end": funcs[n][3]}) for n in sorted(func_names)]
            self.changes.addedTypes = [(n, types[n][1], {"start": types[n][2], "end": types[n][3]}) for n in sorted(type_names)]
            self.changes.addedExternals = [(n, exts[n][1], {"start": exts[n][2], "end": exts[n][3]}) for n in sorted(ext_names)]
        
        return self.changes

File: codetraverse/ast_diff/resdiffer.ts
import { exec } from 'child_process';
import { SyntaxNode } from 'tree-sitter';
import { DetailedChanges } from './Detailedchanges';
import { BaseFileDiff } from './basefilediff';

function formatRescriptFile(filePath: string) {
    try {
        exec(`npx rescript format ${filePath}`);
    } catch (e) {
        // ignore
    }
}

export class RescriptFileDiff extends BaseFileDiff {
    constructor(moduleName: string = "") {
        super(moduleName);
        this.changes = new DetailedChanges(moduleName);
    }

    getDeclName(node: SyntaxNode, nodeType: string | null, nameType: string): string | null {
        for (const child of node.children) {
            if (nodeType && child.type === nodeType) {
                for (const grandchild of child.children) {
                    if (grandchild.isNamed && grandchild.type === nameType) {
                        return grandchild.text;
                    }
                }
            } else if (!nodeType && child.isNamed && child.type === nameType) {
                return child.text;
            }
        }
        return null;
    }

    deepEqual(nodeA: SyntaxNode | null, nodeB: SyntaxNode | null): boolean {
        if (!nodeA || !nodeB) {
            return nodeA === nodeB;
        }

        if (nodeA.type !== nodeB.type) {
            return false;
        }

        const childrenA = nodeA.children;
        const childrenB = nodeB.children;

        if (childrenA.length !== childrenB.length) {
            return false;
        }

        if (childrenA.length === 0) {
            return nodeA.text === nodeB.text && nodeA.parent?.text === nodeB.parent?.text;
        }

        for (let i = 0; i < childrenA.length; i++) {
            if (!this.deepEqual(childrenA[i], childrenB[i])) {
                return false;
            }
        }

        return true;
    }

    extractComponents(root: SyntaxNode): any[] {
        const queue = [root];
        const functions: any = {};
        const types: any = {};
        const externals: any = {};

        const nodeNameMapper: Record<string, [any, (node: SyntaxNode) => string | null]> = {
            "let_declaration": [functions, (x: SyntaxNode) => this.getDeclName(x, "let_binding", "value_identifier")],
            "type_declaration": [types, (x: SyntaxNode) => this.getDeclName(x, "type_binding", "type_identifier")],
            "external_declaration": [externals, (x: SyntaxNode) => this.getDeclName(x, null, "value_identifier")]
        };

        while (queue.length > 0) {
            const currentNode = queue.pop()!;
            if (currentNode.type in nodeNameMapper) {
                const [dct, mapperFunction] = nodeNameMapper[currentNode.type];
                let name = mapperFunction(currentNode);
                if (name) {
                    if (currentNode.parent?.type !== "source_file") {
                        try {
                            name = `${currentNode.parent?.parent?.child(0)?.text}::${name}`;
                        } catch (e) {
                            // ignore
                        }
                    }
                    dct[name] = [currentNode, currentNode.text, currentNode.startPosition, currentNode.endPosition];
                }
            } else {
                for (let i = currentNode.children.length - 1; i >= 0; i--) {
                    const child = currentNode.children[i];
                    if (child.isNamed) {
                        queue.push(child);
                    }
                }
            }
        }
        return [functions, types, externals];
    }

    diffComponents(beforeMap: any, afterMap: any): any {
        const beforeNames = new Set(Object.keys(beforeMap));
        const afterNames = new Set(Object.keys(afterMap));

        const addedNames = Array.from(afterNames).filter(name => !beforeNames.has(name));
        const deletedNames = Array.from(beforeNames).filter(name => !afterNames.has(name));
        const commonNames = Array.from(beforeNames).filter(name => afterNames.has(name));

        const added = addedNames.sort().map(n => [n, afterMap[n][1], { start: afterMap[n][2], end: afterMap[n][3] }]);
        const deleted = deletedNames.sort().map(n => [n, beforeMap[n][1], { start: beforeMap[n][2], end: beforeMap[n][3] }]);

        const modified = [];
        for (const name of commonNames.sort()) {
            const [oldAst, oldBody, oldStart, oldEnd] = beforeMap[name];
            const [newAst, newBody, newStart, newEnd] = afterMap[name];
            if (!this.deepEqual(oldAst, newAst)) {
                modified.push([name, oldBody, newBody, { old_start: oldStart, old_end: oldEnd, new_start: newStart, new_end: newEnd }]);
            }
        }

        return { added, deleted, modified };
    }

    compareTwoFiles(oldFileAst: SyntaxNode, newFileAst: SyntaxNode): DetailedChanges {
        const [oldFuncs, oldTypes, oldExt] = this.extractComponents(oldFileAst);
        const [newFuncs, newTypes, newExt] = this.extractComponents(newFileAst);

        const categoryMap = {
            "functions": [oldFuncs, newFuncs],
            "types": [oldTypes, newTypes],
            "externals": [oldExt, newExt],
        };

        for (const category in categoryMap) {
            const [oldMap, newMap] = categoryMap[category as keyof typeof categoryMap];
            const diff = this.diffComponents(oldMap, newMap);
            for (const changeType in diff) {
                for (const item of diff[changeType]) {
                    this.changes.add_change(category, changeType, item);
                }
            }
        }

        return this.changes;
    }

    processSingleFile(fileAst: SyntaxNode, mode: string = "deleted"): DetailedChanges {
        const [funcs, types, exts] = this.extractComponents(fileAst);
        const funcNames = Object.keys(funcs).sort();
        const typeNames = Object.keys(types).sort();
        const extNames = Object.keys(exts).sort();

        if (mode === "deleted") {
            this.changes.changes['deletedFunctions'] = funcNames.map(n => [n, funcs[n][1], { start: funcs[n][2], end: funcs[n][3] }]);
            this.changes.changes['deletedTypes'] = typeNames.map(n => [n, types[n][1], { start: types[n][2], end: types[n][3] }]);
            this.changes.changes['deletedExternals'] = extNames.map(n => [n, exts[n][1], { start: exts[n][2], end: exts[n][3] }]);
        } else {
            this.changes.changes['addedFunctions'] = funcNames.map(n => [n, funcs[n][1], { start: funcs[n][2], end: funcs[n][3] }]);
            this.changes.changes['addedTypes'] = typeNames.map(n => [n, types[n][1], { start: types[n][2], end: types[n][3] }]);
            this.changes.changes['addedExternals'] = extNames.map(n => [n, exts[n][1], { start: exts[n][2], end: exts[n][3] }]);
        }

        return this.changes;
    }
}
